### PR TITLE
Support for automatically generating backreferences for SDOs

### PR DESCRIPTION
**Pull request by bakkot**
In pursuit of https://github.com/tc39/ecma262/issues/1950, specifically https://github.com/tc39/ecma262/issues/1950#issuecomment-636945295.

Marked as draft until I've tried it out on the full spec, with https://github.com/tc39/ecma262/issues/1950 done.

### CODE REVIEW

File: css/elements.css [hunk 1]
```diff
@@ -278,6 +278,15 @@ emu-rhs {
     text-indent: -25px;
 }
 
+emu-production:not([collapsed]) emu-rhs {
```
Comment (michaelficarra) [inline css/elements.css:281]: #275.

File: css/elements.css [hunk 2]
```diff
@@ -854,20 +863,25 @@ li.menu-search-result-term:before {
...
+  padding-bottom: 7px;
+}
+
+.toolbox-container.active {
+  display: inline-block;
+}
+
 .toolbox {
-	position: absolute;
+  position: relative;
```
Comment (michaelficarra) [inline css/elements.css:877]: Oof this indentation. Maybe send a PR re-indenting this file and rebase on top?

File: js/menu.js [hunk 3]
```diff
@@ -648,23 +651,24 @@ var Toolbox = {
...
   },
 
   activate: function (el, entry, target) {
     if (el === this._activeEl) return;
+    sdoBox.deactivate();
     this.active = true;
     this.entry = entry;
-    this.$container.classList.add('active');
+    this.$outer.classList.add('active');
     this.top = el.offsetTop - this.$container.offsetHeight - 10;
```
Comment (michaelficarra) [inline js/menu.js:663]:
```suggestion
    this.top = el.offsetTop - this.$outer.offsetHeight;
```

**Comment by michaelficarra** [comment]
I don't see `sdoMap.js` in this diff.

File: js/sdoMap.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      sdoBox.activate(this);
+    });
+
+    node.addEventListener('pointerleave', function (e) {
+      sdoBox.deactiveTimeout = setTimeout(function() {
+        if (!insideTooltip) {
+          sdoBox.deactivate();
+        }
+      }, 500);
+      // console.log('exit', this);
```
Comment (michaelficarra) [inline js/sdoMap.js:68]: console.log

File: js/sdoMap.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    insideTooltip = true;
+  });
+  sdoBox.$outer.addEventListener('pointerleave', function () {
+    insideTooltip = false;
+    sdoBox.deactivate();
+  });
+
+  sdoBox.deactiveTimeout = null;
+  [].forEach.call(document.querySelectorAll('emu-grammar[type=definition] emu-rhs'), function (node) {
+    node.addEventListener('pointerenter', function () {
```
Comment (michaelficarra) [inline js/sdoMap.js:58]: I'm not sure whether event delegation would be better for these two listeners. We should ask somebody who does more frontend stuff to advise.

File: js/sdoMap.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+var sdoMap;
+var sdoMapContainer = document.getElementById('sdo-map');
+if (sdoMapContainer == null) {
+  console.error('could not find SDO map container');
+  sdoMap = {};
+} else {
+  sdoMap = JSON.parse(sdoMapContainer.textContent);
+}
+
+console.log(sdoMap)
```
Comment (michaelficarra) [inline js/sdoMap.js:89]: console.log

File: css/elements.css (hunk 2 above)
Comment (bakkot) [reply css/elements.css:877]: https://github.com/tc39/ecmarkup/pull/278

**Review by michaelficarra — APPROVED** [review]

**Comment by bakkot** [comment]
Updated. I uploaded a build of ecma262 made using this PR [here](https://bakkot.github.io/ecma262-previews/77595f5/) so you can see how it looks.

File: js/menu.js [hunk 7]
```diff
@@ -873,7 +873,14 @@ var referencePane = {
 
   showSDOs: function(sdos, alternativeId) {
     this.activate();
-    this.$headerText.textContent = 'Syntax-Directed Operations';
+    var rhs = document.getElementById(alternativeId);
+    var parentName = rhs.parentNode.getAttribute('name')
+    rhs = rhs.cloneNode(true);
+    rhs.querySelectorAll('emu-params,emu-constraints').forEach(function (e) { e.remove(); });
+    rhs.querySelectorAll('emu-opt').forEach(function (e) { e.innerHTML = '_opt' });
```
Comment (michaelficarra) [inline js/menu.js]: I don't like this rendering. I would much prefer these still be rendered as subscripts.

Screenshot 2021-01-23 at 22 52 13

**Comment by michaelficarra** [comment]
I changed the styling of the productions in the menu header. I think it's acceptable this way, but I'd prefer the styling to be the same as in the main body (family, weight, italics). But other than subscripting the optionality mark as I mentioned above, I think that can wait until a follow-up PR.

File: src/Spec.ts [hunk 8]
```diff
@@ -938,15 +981,72 @@ export default class Spec {
...
+              });
+              continue;
+            }
+            let match = matches[0];
+            if (match.id == '') {
+              match.id = 'prod-' + sha(`${name} : ${match.textContent}`);
+            }
+            let mainId = match.id;
+            if (rhs.id == '') {
+              rhs.id = 'prod-' + sha(`[${sdoName}] ${name} ${rhs.textContent}`);
```
Comment (michaelficarra) [inline src/Spec.ts:1037]: You know I'd prefer this to be an hmac. This should be fine, though, if we don't have an hmac function handy.

**Comment by bakkot** [comment]
> I'd prefer the styling to be the same as in the main body

Eh, easy enough. I tightened the spacing slightly (vs the main document) and dropped the yellow coloring from the `opt` subscript, but otherwise it'll match.

Looks like this:

Screen Shot 2021-01-23 at 11 46 09 PM

Or when hovering the production:
Screen Shot 2021-01-23 at 11 46 13 PM

**Comment by bakkot** [comment]
Fixed the italics for terminals:

Screen Shot 2021-01-24 at 9 38 01 AM
Screen Shot 2021-01-24 at 9 38 07 AM

**Review by michaelficarra — APPROVED** [review]

**Pull request merged**